extensions/control-plane/callback/callback-event-dispatcher/src/main/java/org/eclipse/edc/connector/callback/dispatcher/CallbackEventDispatcher.java

```diff
@@ -0,0 +1,73 @@
+/*
+ *  Copyright (c) 2023 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Bayerische Motoren Werke Aktiengesellschaft (BMW AG) - initial API and implementation
+ *
+ */
+
+package org.eclipse.edc.connector.callback.dispatcher;
+
+import org.eclipse.edc.spi.EdcException;
+import org.eclipse.edc.spi.callback.CallbackEventRemoteMessage;
+import org.eclipse.edc.spi.event.Event;
+import org.eclipse.edc.spi.event.EventEnvelope;
+import org.eclipse.edc.spi.event.EventSubscriber;
+import org.eclipse.edc.spi.message.RemoteMessageDispatcherRegistry;
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.eclipse.edc.spi.types.domain.callback.CallbackAddress;
+
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+
+/**
+ * Subscriber for invoking callbacks associated to {@link Event}. If the {@link CallbackAddress#getEvents()} matches
+ * the {@link Event#name()}, the callback is the invoked using a {@link RemoteMessageDispatcherRegistry} with protocol
+ * extracted by {@link CallbackAddress#getUri()}
+ */
+public class CallbackEventDispatcher<T extends Event> implements EventSubscriber<T> {
+    private final RemoteMessageDispatcherRegistry dispatcher;
+    private final boolean transactional;
+    private final Monitor monitor;
+
+    public CallbackEventDispatcher(RemoteMessageDispatcherRegistry dispatcher, boolean transactional, Monitor monitor) {
+        this.dispatcher = dispatcher;
+        this.transactional = transactional;
+        this.monitor = monitor;
+    }
+
+    @Override
+    public void on(EventEnvelope<T> eventEnvelope) {
+        var callbacks = eventEnvelope.getPayload().getCallbackAddresses()
+                .stream().filter(cb -> cb.isTransactional() == transactional)
+                .collect(Collectors.toList());
+
+        var eventName = eventEnvelope.getPayload().name();
+
+        for (var callback : callbacks) {
+            if (matches(eventName, callback)) {
+                try {
+                    dispatcher.send(Object.class, new CallbackEventRemoteMessage<>(callback, eventEnvelope)).get();
```

Comment:
This should be tied to the configured transactional behavior of the callback, i.e. block if it is transactional.

Reply:
Don't we have a default timeout on the `EdcHttpClient` ? 

It's tied to the transactional. If you see the extensions registers two listeners:

- async -> handles notx callbacks
- sync -> handles tx callbacks

and the base logic always `get`, in case we can introduce a  timeout if there isn't one in the `EdcHttpClient`

Reply:
I don't think this `CallbackEventDispatcher` could assume that underneath there's an already configured `EdcHttpClient`, because there should be a completely different dispatching protocol (mq...) so it should have it's own timeout.

regarding the sync/async, I saw that we register two listeners, but I don't see the condition to avoid the thread blocking in the async case, I mean, we're filtering the callbacks but we're treating them in the same way, while the async ones shouldn't block any thread

Reply:
Yeah it will block but only the thread in the executor service when running async. I didn't want to make the code complex since the the listener will not block the current thread but only one Thread of the `EventRouter`.  But in case it should be clear on the `EventSubscriber` API when you can block or not.

Reply:
oh, right, it will be already on another thread, this can be resolved